decisions/websre/0005-web-sre-kubernetes-secrets-pattern.md

```diff
@@ -0,0 +1,40 @@
+
+The goal of this document is to describe the happy path for passing secrets to kubernetes based applications.
+
+## Decision
+
+Store the secrets in the cloud provider secret manager (aws secret manager, gcp secret manager), then use external secrets (https://github.com/external-secrets/kubernetes-external-secrets) to retrieve those secrets and present them to pods as env variables.  The reason this storage method is preferred is that the cloud secret managers seem to be the best secured, most tested, easiest to use place to store secrets. External secrets connects to the secret managers and presents the secret(s) in the method most often recommended for kubernetes; the 12 Factor App pattern of using environment variables.  One other nice benefit of this system is that references to the secrets can be stored in source code, so the connection between the app and the secret storage is explicit, but we're not checking the secret into git anywhere, risking leaking it. 
+
+There are currently a few ways the secrets can be written to the cloud secret store. We have some terraform that writes the secrets directly to the secrets stores, but most secrets are populated manually by operators.  We expect both methods to be used going forward.  In general, we expect SREs/Operators with elevated privileges to manage the secrets.  This ADR intentionally doesn't include scope for devs to write new secrets, or update existing ones.  That problem is as of today, unsolved.
+
+The default assumption for storage patterns in aws secrets store is that you follow the pattern: `<group>/<app>/<environment>/<secretname>` for example `mozmeao/snippets/stage/database_url`  This seems to make writing aws iam access stanzas easy.  Not needing to be rewritten with each new secret added.  But, enough separation to prevent us from accidentally crossing wires.
+
+We are choosing to not use a few other types of secret storage mechanisms, including (but possibly not limited to):
+* git-crypt secret repos + sops, to k8s secrets directly.
+* Sealed secrets (which uses the kubernetes certs as an asymmetric key for local storage, and a crd in the cluster to present the secrets as env variables to the app)
+* SSM Parameter Store (lambdas/codebuilds use this)
+* no new systems to be built, for example hashicorp vault
+
+We expect CI/CD pipelines to use their current secret storage, and not try to connect them to the secret store.  The one exception is the cloud native pipeline solutions (e.g AWS CodeBuild), which should use the secrets manager if possible.  We may end up duplicating secrets, but in general we should try to keep application secrets and pipeline secrets separate, with pipeline secrets being manually added in the right context.
```

Comment:
Why do we expect CI/CD pipelines to be exempty from this process? Are we formally excluding them from scope for now? Supporting reasoning for this exclusion would be nice to have on record, so that when we're working from this ADR, we understand the nature of the exception.

Reply:
I believe its from the statement of "in general we should try to keep application secrets and pipeline secrets separate". At least, in my reading, this makes sense to me in terms of a workflow (CI/CD pipelines' needs) falling adjacent to "kubernetes-based applications" secrets management.

Reply:
That makes sense to me, thanks for clarifying. I don't have any specific change to request in light of that explanation.